contest_service/src/main.rs

```diff
@@ -1,207 +1,296 @@
+                    code: if update_result.matched_count == 0 {
+                        set_user_response::Code::Add as i32
+                    } else {
+                        set_user_response::Code::Update as i32
+                    },
+                })
+            })
     }
-    async fn set_contest(
+    async fn set_contest_metadata(
         &self,
-        _request: Request<SetContestRequest>,
-    ) -> Result<Response<SetContestResponse>, Status> {
-        todo!();
+        request: Request<SetContestMetadataRequest>,
+    ) -> Result<Response<SetContestMetadataResponse>, Status> {
+        let metadata = mappings::contest::ContestMetadata::try_from(request.into_inner())
+            .map_err(|err| Status::invalid_argument(format!("{:?}", err)))?;
+
+        self.get_contest_metadata_collection()
+            .update_one(
+                doc! {},
+                doc! { "$set": Document::from(metadata) },
+                UpdateOptions::builder().upsert(true).build(),
+            )
+            .await
+            .map_err(internal_error)
+            .map(|_| Response::new(SetContestMetadataResponse {}))
     }
     async fn set_problem(
         &self,
-        _request: Request<SetProblemRequest>,
+        request: Request<SetProblemRequest>,
     ) -> Result<Response<SetProblemResponse>, Status> {
-        todo!();
-    }
-    async fn add_question(
-        &self,
-        _request: Request<AddQuestionRequest>,
-    ) -> Result<Response<AddQuestionResponse>, Status> {
-        todo!();
+        let problem_data_from_req = request.into_inner();
+        if let Some(p) = problem_data_from_req.info {
+            if let Some(bin) = problem_data_from_req.statement {
+                let problem_data: mappings::problem::ProblemData = (p.into(), bin).into();
+
+                let document: Document = problem_data.into();
+                self.get_problems_collection()
+                    .update_one(
+                        doc! {"_id": document.get_i32("_id").unwrap()},
+                        doc! { "$set": document },
+                        UpdateOptions::builder().upsert(true).build(),
+                    )
+                    .await
+                    .map_err(internal_error)
+                    .map(|_| Response::new(SetProblemResponse {}))
+            } else {
+                Err(Status::invalid_argument("Missing required parameter"))
+            }
+        } else {
+            Err(Status::invalid_argument("Missing required parameter"))
+        }
     }
-    async fn add_announcement(
+    async fn add_message(
         &self,
-        _request: Request<AddAnnouncementRequest>,
-    ) -> Result<Response<AddAnnouncementResponse>, Status> {
-        todo!();
+        request: Request<AddMessageRequest>,
+    ) -> Result<Response<AddMessageResponse>, Status> {
+        let message = mappings::chat::Message::from(request.into_inner());
+        // TODO should we notify someone here?
```

Comment:
Yes, but not as yet since we need to set up the ParticipantsWebServer and AdminWebServer.

Reply:
Then I'll leave the TODO